### PR TITLE
Async Future Throws RayError as well

### LINKED ISSUE #12384: [Asyncio] await RayActorError throw the wrong error
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
```python
import ray
import asyncio

ray.init()
@ray.remote
class A:
    def hi(self): return "hi"

a = A.remote()
async def main():
    await a.hi.remote()
    ray.kill(a)
    await a.hi.remote()

asyncio.get_event_loop().run_until_complete(main())
```

You will see
```
Exception in callback async_set_result.<locals>.set_future()
handle: <Handle async_set_result.<locals>.set_future()>
Traceback (most recent call last):
  File "/Users/simonmo/miniconda3/lib/python3.6/asyncio/events.py", line 145, in _run
    self._callback(*self._args)
  File "python/ray/_raylet.pyx", line 1530, in ray._raylet.async_set_result.set_future
AttributeError: 'RayActorError' object has no attribute 'as_instanceof_cause'

```
### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR make sure RayError in general is throw directly you call `await ObjectRef`.

## Related issue number
Fixes #12384

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1530,6 +1530,9 @@ cdef void async_set_result(shared_ptr[CRayObject] obj,
         if isinstance(result, RayTaskError):
             ray.worker.last_task_error_raise_time = time.time()
             py_future.set_exception(result.as_instanceof_cause())
+        elif isinstance(result, RayError):
+            # Directly raise exception for RayActorError
+            py_future.set_exception(result)
```
Comment (ijrsvt) [inline python/ray/_raylet.pyx:1535]: Why not also do this for _all_ `Exception`s

**Review by ijrsvt — APPROVED** [review]
Looks good--thanks for catching this :)

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (simon-mo) [reply python/ray/_raylet.pyx:1535]: Hmm this is a good point. The objectref can only be RayError or the object itself. if it's other exception, we should fail loudly here.

**Comment by ijrsvt** [comment]
Oh! regular errors are wrapped in RayError's at this point.

**Pull request merged**